File: src/main/java/de/silencio/activecraftcore/commands/BowCommand.java
package de.silencio.activecraftcore.commands;

import de.silencio.activecraftcore.exceptions.ActiveCraftException;
import de.silencio.activecraftcore.exceptions.InvalidNumberException;
import de.silencio.activecraftcore.utils.ComparisonType;
import org.bukkit.ChatColor;
import org.bukkit.Material;
import org.bukkit.command.Command;
import org.bukkit.command.CommandSender;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.entity.EntityType;
import org.bukkit.entity.Player;
import org.bukkit.entity.Projectile;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.ProjectileHitEvent;
import org.bukkit.event.entity.ProjectileLaunchEvent;
import org.bukkit.inventory.ItemFlag;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.projectiles.ProjectileSource;

import java.util.HashMap;
import java.util.List;

public class BowCommand extends ActiveCraftCommand implements Listener {

    public BowCommand() {
        super("bow");
    }

    private static final HashMap<Projectile, String> projectiles = new HashMap<>();

    @Override
    public void runCommand(CommandSender sender, Command command, String label, String[] args) throws ActiveCraftException {
        checkArgsLength(args, ComparisonType.GREATER_EQUAL, 1);
        switch (args[0].toLowerCase()) {
            case "explode" -> {
                Player player = getPlayer(sender);
                checkPermission(sender, "bow.explode");

                ItemStack boombow = new ItemStack(Material.BOW);
                ItemMeta boombowmeta = boombow.getItemMeta();
                boombowmeta.setDisplayName(ChatColor.GOLD + "Boom Bow");
                boombowmeta.setUnbreakable(true);
                if (args.length > 1)
                    boombowmeta.setLore(List.of(ChatColor.AQUA + "Power: " + parseFloat(args[1])));
                boombow.setItemMeta(boombowmeta);
                boombow.addUnsafeEnchantment(Enchantment.WATER_WORKER, 124);
                boombow.addEnchantment(Enchantment.ARROW_INFINITE, 1);
                boombow.addItemFlags(ItemFlag.HIDE_ENCHANTS);
                player.getInventory().addItem(boombow);
            }
            case "lightning" -> {
                Player player = getPlayer(sender);
                checkPermission(sender, "bow.lightning");

                ItemStack lightningbow = new ItemStack(Material.BOW);
                ItemMeta lightningbowbowmeta = lightningbow.getItemMeta();
                lightningbowbowmeta.setDisplayName(ChatColor.GOLD + "Lightning Bow");
                lightningbowbowmeta.setUnbreakable(true);
                lightningbow.setItemMeta(lightningbowbowmeta);
                lightningbow.addUnsafeEnchantment(Enchantment.WATER_WORKER, 124);
                lightningbow.addEnchantment(Enchantment.ARROW_INFINITE, 1);
                lightningbow.addItemFlags(ItemFlag.HIDE_ENCHANTS);
                player.getInventory().addItem(lightningbow);
            }
        }
    }

    @Override
    public List<String> onTab(CommandSender sender, Command command, String alias, String[] args) {
        return args.length == 1 ? List.of("explode", "lightning") : null;
    }

    @EventHandler
    public void onBowShot(ProjectileLaunchEvent event) {
        Projectile entity = event.getEntity();
        ProjectileSource source = entity.getShooter();

        if (!(source instanceof Player player)) return;
        if (!(entity.getType() == EntityType.ARROW)) return;
        ItemStack item = player.getInventory().getItemInMainHand();

        if (item.getType() != Material.BOW) return;
        ItemMeta itemMeta = item.getItemMeta();

        if (!item.hasItemFlag(ItemFlag.HIDE_ENCHANTS)) return;
        if (!itemMeta.hasEnchant(Enchantment.WATER_WORKER)) return;
        if (!(itemMeta.getEnchantLevel(Enchantment.WATER_WORKER) == 124)) return;

        if (player.hasPermission("activecraft.bow.explode.trigger") && itemMeta.getDisplayName().equalsIgnoreCase(ChatColor.GOLD + "Boom Bow")) {
            float power = 4f;
            List<String> lore = itemMeta.getLore();
            if (lore != null && lore.size() > 0) {
                try {
                    power = parseFloat(lore.get(0).replace(ChatColor.AQUA + "Power: ", ""));
                } catch (InvalidNumberException ignored) {
                }
            }
            projectiles.put(entity, "e" + power);
        }
        if (player.hasPermission("activecraft.bow.lightning.trigger") && itemMeta.getDisplayName().equalsIgnoreCase(ChatColor.GOLD + "Lightning Bow")) {
            projectiles.put(entity, "l");
        }
    }

    @EventHandler(priority = EventPriority.HIGH)
    public void onProjectileHit(ProjectileHitEvent event) {
        Projectile entity = event.getEntity();
        if (!projectiles.containsKey(entity)) return;

        if (projectiles.get(entity).startsWith("e"))
            entity.getWorld().createExplosion(entity.getLocation(), Float.parseFloat(projectiles.get(entity).replace("e", "")), false, true);
        else if (projectiles.get(entity).startsWith("l"))
            entity.getWorld().strikeLightning(entity.getLocation());
        entity.remove();
        projectiles.remove(entity);
    }
}

File: src/main/java/org/activecraft/activecraftcore/commands/BowCommand.kt
package org.activecraft.activecraftcore.commands

import org.activecraft.activecraftcore.ActiveCraftPlugin
import org.activecraft.activecraftcore.exceptions.ActiveCraftException
import org.activecraft.activecraftcore.exceptions.InvalidNumberException
import org.activecraft.activecraftcore.utils.ComparisonType
import org.bukkit.ChatColor
import org.bukkit.Material
import org.bukkit.command.Command
import org.bukkit.command.CommandSender
import org.bukkit.enchantments.Enchantment
import org.bukkit.entity.EntityType
import org.bukkit.entity.Player
import org.bukkit.entity.Projectile
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.entity.ProjectileHitEvent
import org.bukkit.event.entity.ProjectileLaunchEvent
import org.bukkit.inventory.ItemFlag
import org.bukkit.inventory.ItemStack
import java.util.*

class BowCommand(plugin: ActiveCraftPlugin) : ActiveCraftCommand("bow", plugin), Listener {
    @Throws(ActiveCraftException::class)
    public override fun runCommand(sender: CommandSender, command: Command, label: String, args: Array<String>) {
        assertArgsLength(args, ComparisonType.GREATER_EQUAL, 1)
        when (args[0].lowercase()) {
            "explode" -> {
                val player = getPlayer(sender)
                assertCommandPermission(sender, "explode")
                val boombow = ItemStack(Material.BOW)
                val boombowmeta = boombow.itemMeta
                boombowmeta.setDisplayName(ChatColor.GOLD.toString() + "Boom Bow")
                boombowmeta.isUnbreakable = true
                if (args.size > 1) boombowmeta.lore =
                    listOf(ChatColor.AQUA.toString() + "Power: " + parseFloat(args[1]))
                boombow.itemMeta = boombowmeta
                boombow.addUnsafeEnchantment(Enchantment.WATER_WORKER, 124)
                boombow.addEnchantment(Enchantment.ARROW_INFINITE, 1)
                boombow.addItemFlags(ItemFlag.HIDE_ENCHANTS)
                player.inventory.addItem(boombow)
            }

            "lightning" -> {
                val player = getPlayer(sender)
                assertCommandPermission(sender, "lightning")
                val lightningbow = ItemStack(Material.BOW)
                val lightningbowbowmeta = lightningbow.itemMeta
                lightningbowbowmeta.setDisplayName(ChatColor.GOLD.toString() + "Lightning Bow")
                lightningbowbowmeta.isUnbreakable = true
                lightningbow.itemMeta = lightningbowbowmeta
                lightningbow.addUnsafeEnchantment(Enchantment.WATER_WORKER, 124)
                lightningbow.addEnchantment(Enchantment.ARROW_INFINITE, 1)
                lightningbow.addItemFlags(ItemFlag.HIDE_ENCHANTS)
                player.inventory.addItem(lightningbow)
            }
        }
    }

    public override fun onTab(
        sender: CommandSender,
        command: Command,
        alias: String,
        args: Array<String>
    ): List<String>? {
        return if (args.size == 1) listOf("explode", "lightning") else null
    }

    @EventHandler
    fun onBowShot(event: ProjectileLaunchEvent) {
        val entity = event.entity
        val source = entity.shooter as? Player ?: return
        if (entity.type != EntityType.ARROW) return
        val item = source.inventory.itemInMainHand
        if (item.type != Material.BOW) return
        val itemMeta = item.itemMeta
        if (!item.hasItemFlag(ItemFlag.HIDE_ENCHANTS)) return
        if (!itemMeta.hasEnchant(Enchantment.WATER_WORKER)) return
        if (itemMeta.getEnchantLevel(Enchantment.WATER_WORKER) != 124) return
        if (source.hasPermission("activecraft.bow.explode.trigger") && itemMeta.displayName.equals(
                ChatColor.GOLD.toString() + "Boom Bow",
                ignoreCase = true
            )
        ) {
            var power = 4f
            val lore = itemMeta.lore
            if (lore != null && lore.size > 0) {
                try {
                    power = parseFloat(lore[0].replace(ChatColor.AQUA.toString() + "Power: ", ""))
                } catch (ignored: InvalidNumberException) {
                }
            }
            projectiles[entity] = "e$power"
        }
        if (source.hasPermission("activecraft.bow.lightning.trigger") && itemMeta.displayName.equals(
                ChatColor.GOLD.toString() + "Lightning Bow",
                ignoreCase = true
            )
        ) {
            projectiles[entity] = "l"
        }
    }

    @EventHandler(priority = EventPriority.HIGH)
    fun onProjectileHit(event: ProjectileHitEvent) {
        val entity = event.entity
        if (!projectiles.containsKey(entity)) return
        if (projectiles[entity]!!.startsWith("e")) entity.world.createExplosion(
            entity.location, projectiles[entity]!!
                .replace("e", "").toFloat(), false, true
        ) else if (projectiles[entity]!!.startsWith("l")) entity.world.strikeLightning(entity.location)
        entity.remove()
        projectiles.remove(entity)
    }

    companion object {
        private val projectiles: MutableMap<Projectile, String> = mutableMapOf()
    }
}